File: Koshelev/1609.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool BFS(vector<TreeNode*> LevelNodes, int level)
    {
        if (LevelNodes.size() == 0)
            return true;
        else
        {
            if (LevelNodes.size() > 1)
            {
                for (int i = 0; i < LevelNodes.size() - 1; i++)
                {
                    if (level % 2 == 0)
                    {
                        if (LevelNodes[i]->val >= LevelNodes[i + 1]->val || LevelNodes[i]->val % 2 == 0)
                            return false;
                    }
                    else if (LevelNodes[i]->val <= LevelNodes[i + 1]->val || LevelNodes[i]->val % 2 == 1)
                        return false;
                }
            }

            if (LevelNodes[LevelNodes.size() - 1]->val % 2 == level % 2)
                return false;

            vector<TreeNode*> NewLevelNodes;

            for (int i = 0; i < LevelNodes.size(); i++)
            {
                if (LevelNodes[i]->left != nullptr)
                    NewLevelNodes.push_back(LevelNodes[i]->left);
                if (LevelNodes[i]->right != nullptr)
                    NewLevelNodes.push_back(LevelNodes[i]->right);
            }

            return BFS(NewLevelNodes, level + 1);
        }
    }

    bool isEvenOddTree(TreeNode* root) {
        vector<TreeNode*> LevelNode;
        LevelNode.push_back(root);
        return BFS(LevelNode, 0);                
    }
};

File: 1609.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def isEvenOddTree(self, root):
        if not root:
            return True

        queue = [root]
        level = 0

        while queue:
            level_vals = []
            size = len(queue)

            for _ in range(size):
                node = queue.pop(0)
                val = node.val

                if level % 2 == 0:
                    if val % 2 == 0 or (level_vals and val <= level_vals[-1]):
                        return False
                else:
                    if val % 2 == 1 or (level_vals and val >= level_vals[-1]):
                        return False

                level_vals.append(val)

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            level += 1

        return True